src/Lecture2.hs

```diff
@@ -36,260 +42,145 @@ module Lecture2
-       treasure besides gold (if you already haven't done this).
--}
+dropSpaces :: String -> String
+dropSpaces str = takeWhile (not.isSpace) $ dropWhile isSpace str
 
--- some help in the beginning ;)
 data Knight = Knight
     { knightHealth    :: Int
     , knightAttack    :: Int
     , knightEndurance :: Int
     }
 
-dragonFight = error "TODO"
+data DragonType = Red| Black |Green
 
-----------------------------------------------------------------------------
--- Extra Challenges
-----------------------------------------------------------------------------
+data Dragon = Dragon
+    { dragonType :: DragonType
+	, dragonHealth :: Int
+	, dragonFirePower :: Int
+	, treasureChest :: Maybe TreasureChest 
+	, expPoints :: Int}
 
-{- The following exercises are considered optional. Some of them might be more
-challenging. However, you still may find some of them easier than some of the
-previous ones. Difficulty is a relative concept.
--}
-
-{- | Write a function that takes a list of numbers and returns 'True'
-if all the numbers are in the increasing order (i.e. the list is
-sorted).
+data TreasureChest = TreasureChest
+    { gold :: Int
+	, treasure :: Int}
 
->>> isIncreasing [3, 1, 2]
-False
->>> isIncreasing [1 .. 10]
-True
--}
-isIncreasing :: [Int] -> Bool
-isIncreasing = error "TODO"
+data FightOutcome = DragonDead | KnightDead | KnightRunsAway | DragonAlive
 
-{- | Implement a function that takes two lists, sorted in the
-increasing order, and merges them into new list, also sorted in the
-increasing order.
+dragonFight:: Dragon -> Knight -> FightOutcome
+dragonFight d k = let dH = calculateDragonHealth (dragonHealth d) (knightAttack k)
+                      kH = calculateKnightHealth (dragonFirePower d) (knightHealth k)
+                      kE = calculateKnightEndurance (knightAttack k) (knightEndurance k)
+				  in case (dH,kH,kE) of
+				         (dH, _,_) | dH<= 0-> DragonDead
+				         (dH, _,_) | dH> 0-> DragonAlive
+				         (_,kH,_)  |kH<=0-> KnightDead
+				         (_,_,kE)  |kE==0 -> KnightRunsAway
 
-The lists are guaranteed to be given sorted, so you don't need to
-verify that.
+calculateDragonHealth::Int -> Int -> Int
+calculateDragonHealth dragonHealth knightAttack = dragonHealth - knightAttack
 
->>> merge [1, 2, 4] [3, 7]
-[1,2,3,4,7]
--}
-merge :: [Int] -> [Int] -> [Int]
-merge = error "TODO"
+calculateKnightHealth::Int -> Int -> Int
+calculateKnightHealth dragonFirePower knightHealth = knightHealth - dragonFirePower
 
-{- | Implement the "Merge Sort" algorithm in Haskell. The @mergeSort@
-function takes a list of numbers and returns a new list containing the
-same numbers but in the increasing order.
+calculateKnightEndurance::Int -> Int -> Int
+calculateKnightEndurance knightAttack knightEndurance = knightEndurance - knightAttack
 
-The algorithm of merge sort is the following:
+isIncreasing :: [Int] -> Bool
+isIncreasing [] = True
+isIncreasing [x] = True
```

Comment:
You can replace a variable name with an underscore here since you don't use the variable
```suggestion
isIncreasing [_] = True
```